File: src/solutions/2025/05/csharp/Main.cs

using System.ComponentModel.DataAnnotations;
using System.Diagnostics;
using Core;

namespace Year2025;

public static class Day05
{
   static SolutionOptions OPTIONS = new()
   {
      HasAlternate = false,
      HasIO = false,
   };

   static private string Solve(string input, bool isTest, bool p2)
   {
      var chunks = input.Split("\n\n");
      var ranges = chunks[0]
         .Split('\n')
         .Select(x =>
         {
            var m = x.Split('-')
               .Select(long.Parse)
               .ToArray();
            return (min: m[0], max: m[1]);
         })
         .ToArray();

      if (p2)
      {
         var highestMin = 0L;
         return ranges
            .OrderBy(r => r.min)
            .Sum(current =>
            {
               var res = Math.Max(current.max + 1 - Math.Max(current.min, highestMin), 0);
               highestMin = Math.Max(highestMin, current.max + 1);
               return res;
            })
            .ToString();
      }
      return chunks[1]
         .Split('\n')
         .Select(long.Parse)
         .Count(x => ranges.Any(r => x >= r.min && x <= r.max))
         .ToString();

   }

   static string Part1(string input, bool isTest = false)
   {
      return Solve(input, isTest, false);
   }

   static string Part2(string input, bool isTest = false)
   {
      return Solve(input, isTest, true);
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}

File: src/solutions/2025/05/ts/main.ts
import { argv } from "node:process";
import type { SolutionOptions } from "src/options.ts";
import { run } from "src/run.ts";

const options: SolutionOptions = {
   hasAlternate: false,
   hasIo: false,
};

function solve(input: string, isTest: boolean, p2: boolean): string {
   const chunks = input.split("\n\n");
   const ranges = chunks[0].split("\n").map((x) => x.split("-").map(Number));

   if (p2) {
      let highestMin = 0;
      return ranges
         .sort((a, b) => a[0] - b[0])
         .reduce((acc, current) => {
            const res = Math.max(
               current[1] + 1 - Math.max(current[0], highestMin),
               0,
            );
            highestMin = Math.max(highestMin, current[1] + 1);
            return acc + res;
         }, 0)
         .toString();
   }
   return chunks[1]
      .split("\n")
      .map(Number)
      .filter((x) => ranges.some((r) => r[0] <= x && x <= r[1]))
      .length.toString();
}

function part1(input: string, isTest: boolean): string {
   return solve(input, isTest, false);
}

function part2(input: string, isTest: boolean): string {
   return solve(input, isTest, true);
}

if (import.meta.main) {
   run(argv, part1, part2, options);
}
